### PR TITLE
client: support getfattr ceph.dir.pin extended attribute

### PR DESCRIPTION
**Pull request by david-z**
In Multi-MDSes, we can set ceph.dir.pin on client to bind a directory to a specific MDS. But we can't get this attribute on client. So it is not convenient to check which directory is pinned on which MDS since we have multiple active MDSes running in a cluster and many user direcotires pinned to different MDS.

Fixes: http://tracker.ceph.com/issues/36707

Signed-off-by: Zhi Zhang <zhangz.david@outlook.com>

- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by david-z** [comment]
@batrick @ukernel 

1. I didn't add a new CEPH FEATURE BIT for this feature, because I feel this is just a tiny feature and no need to occupy one CEPH FEATURE BIT. I used a new flag for VXattr, like we introduced VXATTR_RSTAT before.

2. For CEPH_STAT_DIR_PIN, I defined it as (CEPH_CAP_FILE_WREXTEND << 2). This is the minimum and unsed bit for getattr masks. Please check if it is suitable.

**Comment by david-z** [comment]
retest this please

File: src/mds/Server.cc [hunk 1]
```diff
@@ -3450,6 +3450,10 @@ void Server::handle_client_getattr(MDRequestRef& mdr, bool is_lookup)
 
   // reply
   dout(10) << "reply to stat on " << *req << dendl;
+  if (mask & CEPH_STAT_DIR_PIN) {
+    dout(20) << " encode inode export_pin=" << ref->inode.export_pin << dendl;
+    encode(ref->inode.export_pin, mdr->reply_extra_bl);
```
Comment (batrick) [inline src/mds/Server.cc]: This should go in `CInode::encode_inodestat` I think. Then just increment the encoded stat version.

Comment (david-z) [reply src/mds/Server.cc]: I thought of this at the beginning, but this required to add a CEPH FEATURE BIT and check client session do has this feature before encoding export_pin value. So I switched to use a flag for VXattr.

Comment (david-z) [reply src/mds/Server.cc]: @batrick  Oh, I see what you mean now. The latest master branch is so different from Luminous version at `CInode::encode_inodestat`. I will make the changes accordingly.

**Comment by david-z** [comment]
@batrick  I have run tests based on 14.0.1 on our env. This change works fine. Pls review this PR when you have time.

**Comment by david-z** [comment]
@batrick  Tests have been added. And fix an issue if there is no standby MDS available, following case would always fail.

self._wait_subtrees(status, 0, [('/1', 0), ('/1/4/5', 1), **('/1/2/3', 2)**, ('/a', 1), ('/aa/bb', 0)])

File: src/client/Inode.h [hunk 2]
```diff
@@ -278,7 +280,7 @@ struct Inode {
       snaprealm(0), snaprealm_item(this),
       oset((void *)this, newlayout->pool_id, this->ino),
       reported_size(0), wanted_max_size(0), requested_max_size(0),
-      _ref(0), ll_ref(0)
+      _ref(0), ll_ref(0), dir_pin(-ENODATA)
```
Comment (ukernel) [inline src/client/Inode.h]: use MDS_RANK_NONE here

Comment (david-z) [reply src/client/Inode.h]: @ukernel  I use -ENODATA here for the compatibility consideration. 

1. New ceph-fuse, but old MDS:

Ceph-fuse won't get any pin value from old MDS, so `Client::_vxattrcb_dir_pin_exists` can tell getfattr ceph.dir.pin is not supported. So getfattr will return "No such attribute" like before.

2. New ceph-fuse and new MDS, but directory doesn't set pin value:

Ceph-fuse can decode pin value which is -1 and `Client::_vxattrcb_dir_pin_exists` can tell getfattr ceph.dir.pin is already supported. So getfattr will return -1.

If use MDS_RANK_NONE, we can't distinguish above two cases. What do you think?

Comment (david-z) [reply src/client/Inode.h]: @ukernel  pls ignore above comment. Your suggestion is good.

**Comment by david-z** [comment]
retest this please

**Comment by david-z** [comment]
@batrick @ukernel  Pls review the latest change. And I triggered jenkins build 2 times and they were all failed, but failures are not related to this change.

File: src/messages/MClientReply.h [hunk 3]
```diff
@@ -183,6 +185,11 @@ struct InodeStat {
       decode(layout.pool_ns, p);
       decode(btime, p);
       decode(change_attr, p);
+      if (struct_v > 1) {
+        decode(dir_pin, p);
+      } else {
+        dir_pin = -ENODATA;
```
Comment (ukernel) [inline src/messages/MClientReply.h]: RANK_NONE

File: src/client/Client.cc [hunk 4]
```diff
@@ -11742,6 +11743,14 @@ size_t Client::_vxattrcb_dir_rctime(Inode *in, char *val, size_t size)
   return snprintf(val, size, "%ld.09%ld", (long)in->rstat.rctime.sec(),
       (long)in->rstat.rctime.nsec());
 }
+bool Client::_vxattrcb_dir_pin_exists(Inode *in)
+{
+  return in->dir_pin != -ENODATA;
```
Comment (ukernel) [inline src/client/Client.cc]: RANK_NONE

Comment (david-z) [reply src/client/Client.cc]: not found RANK_NONE? you mean MDS_RANK_NONE?

Comment (david-z) [reply src/client/Client.cc]: If we use MDS_RANK_NONE at both places, and even if ceph-fuse and mds are all new version, but directory is not set any pin value,  getfattr ceph.dir.pin will always return "No such attribute". 

If we use ENODATA at both places,  new version of ceph-fuse and mds will return -1, but new version of ceph-fuse and old version of mds will still return "No such attribute".

Comment (david-z) [reply src/client/Client.cc]: @ukernel  Pls check my comments again, thanks.

**Comment by ukernel** [comment]
LGTM

**Comment by david-z** [comment]
@batrick Pls take a look at this PR. All the required changes have been done. Thanks.

**Comment by david-z** [comment]
retest this please

**Pull request merged**